src/relations/postgresql.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Define the Temporal worker postgresql relation."""
+
+import logging
+
+from ops import framework
+from ops.model import WaitingStatus
+
+from log import log_event_handler
+
+logger = logging.getLogger(__name__)
+
+
+class Postgresql(framework.Object):
+    """Client for postgresql relations."""
+
+    def __init__(self, charm):
+        """Construct.
+
+        Args:
+            charm: The charm to attach the hooks to.
+        """
+        super().__init__(charm, "database")
+        self.charm = charm
+
+        charm.framework.observe(charm.database.on.database_created, self._on_database_changed)
+        charm.framework.observe(charm.database.on.endpoints_changed, self._on_database_changed)
+        charm.framework.observe(charm.on.database_relation_broken, self._on_database_relation_broken)
+
+    @log_event_handler(logger)
+    def _on_database_changed(self, event) -> None:
+        """Handle database creation/change events.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm._state.is_ready():
+            event.defer()
+            return
+
+        self.charm.unit.status = WaitingStatus(f"handling {event.relation.name} change")
+
+        self.update_db_relation_data_in_state(event)
+        self.charm._update(event)
```

Comment:
We talked about (emitting custom events instead of calling) but I am not going to push it if you want to keep it like this.